src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +386,92 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
                         }
                     });
 
-                    var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-                    var codeCleanupService = document.GetLanguageService<ICodeCleanupService>();
+                    var solution = await applyFixAsync(progressTracker, cancellationToken).ConfigureAwait(true);
 
-                    var allDiagnostics = codeCleanupService.GetAllDiagnostics();
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
 
-                    var enabedDiagnosticSets = ArrayBuilder<DiagnosticSet>.GetInstance();
+                    return workspace.TryApplyChanges(solution, progressTracker);
+                }
+            }
+        }
 
-                    foreach (var diagnostic in allDiagnostics.Diagnostics)
-                    {
-                        foreach (var diagnosticId in diagnostic.DiagnosticIds)
-                        {
-                            if (context.EnabledFixIds.IsFixIdEnabled(diagnosticId))
-                            {
-                                enabedDiagnosticSets.Add(diagnostic);
-                                break;
-                            }
-                        }
-                    }
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            progressTracker.AddItems(solution.ProjectIds.Count);
+            foreach (var projectId in solution.ProjectIds)
```

Comment:
>foreach [](start = 12, length = 7)

Can we execute this in parallel? Do we have any existing fixers that fix things across project boundary?

Reply:
➡️ I'm keeping the initial implementation sequential so we can focus on correct behavior. It isn't fast for large solutions, but it does support cancellation if it's taking too long.